### ISSUE TITLE
Write man page

### ISSUE THREAD

**Issue by maltfield** [BusKill/buskill-app#48]
This ticket will track the effort to write a manpage for the BusKill app that will be displayed when running `man buskill` on *nix systems.

This is a prerequisite to getting BusKill added to the official Debian repos in #31

 * https://github.com/BusKill/buskill-app/issues/31

**Comment by maltfield** [comment]
The example given by @fmarier was this repo they hand-wrote for the `pip-check-reqs` package:

 * https://salsa.debian.org/debian/pip-check-reqs/-/blob/master/debian/pip-missing-reqs.1

The format it's written-in is called "Nroff"

 * https://en.wikipedia.org/wiki/Nroff

The GNU tool for parsing nroff files is "Groff"

**Comment by maltfield** [comment]
Thanks :) I'm currently just trying to figure out the best-practice of:

1. Where to store the manpage in this repo
2. Whether to store it in this repo in [a] native-manpage's groff format or [b] reStructuredText like the rest of this project's documentation (and convert to groff format at build-time) or [c] some other pythonic best-practice for dynamically producing the manpage at build time

**Comment by maltfield** [comment]
I spent some time looking at other projects to see how they build and where they store their man pages

* ansible: appears to be generated dynamically with some jinja2 FileSystemLoader? https://github.com/ansible/ansible/blob/2cbfd1e350cbe1ca195d33306b5a9628667ddda8/hacking/build_library/build_ansible/command_plugins/generate_man.py
* cryptsetup has it in '/man/cryptsetup.8.adoc' https://gitlab.com/cryptsetup/cryptsetup/-/blob/main/man/cryptsetup.8.adoc
* flatpak: looks like the man pages are stored in 'doc/' under some .xml format named DocBook and then somehow converted to manpage format by ${AM_V_GEN} ${XSLTPROC} https://github.com/flatpak/flatpak/tree/main/doc
		1. looks like this is XSTL Processor https://developer.mozilla.org/en-US/docs/Web/API/XSLTProcessor
* kdenvlive has a groff-type file at data/man/kdenlive.1. At the top, a comment says the file was generated by kdemangen.pl https://github.com/KDE/kdenlive/blob/master/data/man/kdenlive.1
		1. https://github.com/KDE/kde-dev-scripts/blob/master/kdemangen.pl
		2. they used-to have a file named '/man/kdenlive.1', but it no longer exists https://github.com/KDE/kdenlive/blob/da670ef2418d0894fad0ef22b6908d69d7c19409/man/kdenlive.1
		3. apparently it was moved to data/man/kdenlive.1 in 2014 https://github.com/KDE/kdenlive/commit/5a7add266afeb0ba61580379466451b7ebdf7399
* nmap has a manpage in 'docs/nmap.1' with comments at the top indicating it was generated by "DocBook XSL Stylesheets v1.79.1" https://github.com/nmap/nmap/blob/master/docs/nmap.1
* rclone has a file named 'rclone.1' in the root of its repo, which has a comment at the top indicating it was autogenerated by pandoc https://raw.githubusercontent.com/rclone/rclone/master/rclone.1
		1. this is done in the Makefile https://github.com/rclone/rclone/blob/ce168ecac214c1d729cff0ce776e33b8ef978e13/Makefile#L123
		2. it looks like the source is MANUAL.md https://raw.githubusercontent.com/rclone/rclone/master/MANUAL.md
* VLC stores their manpage in 'docs/vlc.1' and at first-glance it does not appear to be autogenerated by some other tool https://github.com/videolan/vlc/blob/master/doc/vlc.1
* zim has this script that does the bare-miniumum and just tells the user to go to their website, pulling the info dynamically from the github README.md and the python 'zim' class https://github.com/zim-desktop-wiki/zim-desktop-wiki/blob/1ccaffc181671a71b6d66cbe99ae14688db1fc6c/makeman.py
		1. curiously, that differes from the actual `man zim` page that I have on my system

**Comment by maltfield** [comment]
I think the best short-term solution is to manually write our own man page in a `.rst` file somewhere in our `docs/` directory. That way:

1. We can include the man page in our sphinx documentation, similar to how we do with `LICENSE`
2. We can use `pandoc` to spit-out the man-page in groff format, while keeping all this project's docs in rst format

Long-term, at some point in the future, we could probably autogenerate sections of the man page directly from our sourcecode, such as the arguments section. But I'll skip that for now..

**Comment by maltfield** [comment]
Here's some examples of some `.rst` (source files written in reStructuredText) that are used to generate manpages in groff-format

1. https://gist.github.com/rlaager/0881883d66b45fae9cc145fa92203c69
2. https://docutils.sourceforge.io/sandbox/manpage-writer/rst2man.txt

**Comment by maltfield** [comment]
See also https://en.wikipedia.org/wiki/Comparison_of_documentation_generators

**Comment by maltfield** [comment]
See also https://tex.stackexchange.com/questions/668065/missing-th-section-when-converting-rst-file-to-groff-format-in-pandoc-restr

**Comment by maltfield** [comment]
I created a new branch named `manpage` where this work is taking place

**Comment by fmarier** [comment]
This is some pretty in-depth research! It would almost be worth putting into a blog post so that it doesn't disappear into the issue tracker :)

**Comment by maltfield** [comment]
I managed to fix the issue where `pandoc` spat-out the groff file without the `.TH` section by merely adding the `-s` = `--standalone` to its execution. Now I get the desired start/end on the first-line, but it's still missing the short description in the middle for some reason.

```
user@buskill:~$ head -n10 sandbox/buskill-app/docs/software_usr/manpage.rst
.. _manpage:

============
 buskill(1)
============

-------------------------------------------------------------------------
Lock or shutdown if your device is stolen (for use with laptop kill cord)
-------------------------------------------------------------------------

user@buskill:~$ 
...
user@buskill:~$ pandoc -s sandbox/buskill-app/docs/software_usr/manpage.rst -t man | groffer --text | head -n10
buskill(1)                                                          buskill(1)



SYNOPSIS
              buskill --trigger soft-shutdown --arm

DESCRIPTION
       buskill  is  a cross-platform CLI and GUI app that can lock or shutdown
       your computer when a USB hotplug removal event occurs.
user@buskill:~$ 
```

**Comment by maltfield** [comment]
From the pandoc manual, I found that you could ouput the "template" for the `man` output as follows

```
user@buskill:~$ pandoc -D man
$if(has-tables)$
.\"t
$endif$
$if(pandoc-version)$
.\" Automatically generated by Pandoc $pandoc-version$
.\"
$endif$
$if(adjusting)$
.ad $adjusting$
$endif$
.TH "$title/nowrap$" "$section/nowrap$" "$date/nowrap$" "$footer/nowrap$" "$header/nowrap$"
$if(hyphenate)$
.hy
$else$
.nh
$endif$
$for(header-includes)$
$header-includes$
$endfor$
$for(include-before)$
$include-before$
$endfor$
$body$
$for(include-after)$
$include-after$
$endfor$
$if(author)$
.SH AUTHORS
$for(author)$$author$$sep$; $endfor$.
$endif$
user@buskill:~$ 
```

**Comment by maltfield** [comment]
To add all the variable to populate the manpage template via `pandoc`, I had to specify some variables directly on the command-line when doing the conversion between the `.rst` file to the groff file.

So I hacked-up a simple script to wrap this command:

 * https://github.com/BusKill/buskill-app/blob/5efab6ae82d472115109111278dfa5c9c6164df3/docs/buildManpage.sh

**Comment by maltfield** [comment]
I've merged the `manpage` branch into `dev`, so it's now possible to create this project's manpage with the following commands

```
git clone https://github.com/BusKill/buskill-app.git
cd buskill-app
git checkout dev
docs/buildManpage.sh
man ./docs/buskill.1
```

I've confirmed this to be working on Debian 11. Here's an example execution

```
TODO
```

**Comment by maltfield** [comment]
@fmarier can you confirm this man page meet's Debian's needs?

**Comment by fmarier** [comment]
@maltfield That looks good to me.

I filed https://github.com/BusKill/buskill-app/pull/50 for a few nits, but even without it would be good.

**Comment by maltfield** [comment]
Merged PR #50. I'm marking this ticket as complete, and now #31 is unblocked.

**Comment by fmarier** [comment]
Just to add to your list of manpage resources, here's a very good guide I had forgotten about: https://liw.fi/manpages/
